### PR TITLE
Python: add BoundMethodValue (round 2)

### PR DESCRIPTION
**Pull request by RasmusWL**
Fixes github/codeql-python-team#64, 

Before this PR is ready to be merged, we need to solve 2 things:

- [x] 1. `BoundMethodValue.getParameter(-1)`
- [x] 2. `getNamedArgumentForCall` for builtins

---

### 1. `BoundMethodValue.getParameter(-1)`
Is it acceptable that `BoundMethodValue.getParameter(-1)` returns the NameNode for `self`?

Intuitively I just want to disallow it (since it is very surprising and strange), but I have this nagging feeling that we're removing the ability to find the `self` parameter and that this might be useful.

### 2. `getNamedArgumentForCall` for builtins

Why is `getNamedArgumentForCall` restricted to only work for non builtins?

instead of having just
```codeql
call.getArgByName(name) = result
```

it has [this](https://github.com/RasmusWL/ql/blob/dfe7c8270b6ee99ed70fd10c8fadcdf4fe46ea99/python/ql/src/semmle/python/objects/ObjectAPI.qll#L423-L424):

```codeql
call.getArgByName(name) = result and
exists(this.getParameterByName(name))
```

effectively this means that there are no results for `<builtin function or method>.getNamedArgumentForCall(_, _)`.

But why? Looking in the archives, I could not find an explanation (I did not look further than when the ql repo was moved made public). I could come up with these ideas:

* We had an assumption that builtin functions and methods **never** take keyword arguments.
    - This is wrong, since `open` takes keyword arguments, but is a `BuiltinFunctionValue`. (all other builtin functions I investigated did not take keyword arguments though).

* Since we didn't have any mechanism to record the positional index of a keyword argument, we wouldn't be able to offer the translation between positional/keyword arguments. Since that is _the only_ reason you would use `getNamedArgumentForCall` anyway, we chose to make it fail so users would not get any false assumptions.

* We don't want to include arguments that are passed to a `**kwargs`.
    - in which case, we should probably change `getArgumentForCall` so it doesn't give results for `*args`

#### What to do now?

1. I suggest we remove the `exists(this.getParameterByName(name))` restriction.
    - For the call `open(file="foo.txt")` I think it's very unuseful (and surprising) that `getNamedArgumentForCall(call, "file")` doesn't give any result.
    - This will change our behavior in regards to `**kwargs`, which I'm happy about. With `def foo(a, b, **kwargs)`, and `foo(a=1, b=2, c=3)` I think it's very unuseful that `getNamedArgumentForCall(call, "c")` will not have any result :confused: 

2. Long term, I suggest we record the signature of builtin functions and methods, so we can properly support `getArgumentForCall` and `getNamedArgumentForCall`. My impression is that there are only a few builtin functions and methods that take keyword arguments, but I think it's disappointing that we don't properly support those :neutral_face:

### CODE REVIEW

**Comment by tausbn** [comment]
Regarding your first question (having not looked at the actual code yet), I would be in favour of not having "magic" behaviour like the `self` parameter being at index `-1`.
If we need the `self` parameter, perhaps we should just have a `getSelfParameter` method?

**Comment by RasmusWL** [comment]
Done :+1:

File: python/ql/src/semmle/python/objects/Callables.qll [hunk 1]
```diff
@@ -438,16 +438,26 @@ class BoundMethodObjectInternal extends CallableObjectInternal, TBoundMethod {
...
+    override NameNode getParameter(int n) {
+        result = this.getFunction().getParameter(n + 1) and
+        // don't return the parameter for `self` at `n = -1`
+        n >= 0
+    }
 
     override NameNode getParameterByName(string name) {
-        result = this.getFunction().getParameterByName(name)
+        result = this.getFunction().getParameterByName(name) and
+        not result = this.getSelfParameter()
```
[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/semmle/python/objects/Callables.qll:452]: Actually, I take that back. After all, people might prefer to write `def foo(this, self)` so we can't rely on `self` being the correct name for the self parameter.

[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/semmle/python/objects/Callables.qll:452]: Perhaps we should instead extend the QLdoc to mention that if you want the self parameter, you have to use `getSelfParameter`?

File: python/ql/src/semmle/python/objects/ObjectAPI.qll [hunk 2]
```diff
@@ -396,6 +436,29 @@ class CallableValue extends Value {
...
 
+/**
+ * Class representing bound-methods, such as `o.func`, where `o` is an instance
+ * of a class that has a callable attribute `func`.
+ */
+class BoundMethodValue extends CallableValue {
+    BoundMethodValue() { this instanceof BoundMethodObjectInternal }
+
+    /**
+     * Gets the callable that will be used when `this` called.
```
Comment (tausbn) [inline python/ql/src/semmle/python/objects/ObjectAPI.qll:447]:
```suggestion
     * Gets the callable that will be used when `this` is called.
```

File: python/ql/src/semmle/python/objects/Callables.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/semmle/python/objects/Callables.qll:458]: Done :+1:

**Comment by RasmusWL** [comment]
Right, I implemented (1) from _What to do now?_ (in PR description), and created internal tracking ticket for (2). So this should be ready to go :rocket:

**Pull request merged**